### PR TITLE
Refactor and expand validator preconditions to espresso STF logic

### PR DESCRIPTION
**Pull request by zacshowa**
Closes [#140](https://github.com/EspressoSystems/nitro-espresso-integration/issues/140) 

### This PR:

Refactors the current checks before entering the espresso logic for the validator STF into it's own function to improve readability. Additionally, this pr adds an additional item to validate that if the hotshot height is non zero, that the message being received is an espresso message. In the event that the hotshot height is nonzero, and the message received is not an espresso message, the validator now panics.

### This PR does not:
Add a unit test in espresso_machine_test.go as the new panics only occur whenever the message being received is not an espresso message, and I was unable to determine how to feed a non espresso message to the validator logic. 

### Key places to review:

The new file that lives in `nitro-espresso-integration/cmd/replay/espresso_validation.go`

The changed logic on line 283 of `nitro-espresso-integration/cmd/replay/main`

 ### How to test this PR:

`go test -run ^TestEspressoArbMachine$ github.com/offchainlabs/nitro/system_tests -v`  
This test will validate that no previous behavior has been broken by these changes.

### CODE REVIEW

File: cmd/replay/espresso_validation.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// /
+// / Panics: This function will panic if the message type is not an espresso message, but the HotShot height is non zero
+// / 		as this is an invalid state for the STF to reside in.
+// /
+func handleEspressoPreConditions(message *arbostypes.MessageWithMetadata, isEnabled bool) bool {
+	// calculate and cache all values needed to determine if the preconditions are met to enter the Espresso STF logic
+	isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
+	isEspressoMessage := !isNonEspressoMessage
+	hotShotHeight := wavmio.GetEspressoHeight()
+	validatingEspressoLivenessFailure := isNonEspressoMessage && (isEnabled || hotShotHeight != 0)
```
Comment (ImJeremyHe) [inline cmd/replay/espresso_validation.go:27]: `wavmio.GetEspressoHeight` will return the current L2 hotshot height. Not every messages in the espresso mode can be called `EspressoMessage`. EspressoMessage is the `L2Message` (which is for creating blocks and consumes 1 hotshot block) in the espresso mode.
There are some other messages and I believe they also have the hotshot height which is equal to the last EspressoMessage

**Review by ImJeremyHe — COMMENTED** [review]
Does the staker test pass locally?
https://github.com/EspressoSystems/nitro-espresso-integration/blob/e34c95d008cc1e85aeb03c48d8abfaacd3c6f39c/system_tests/espresso_e2e_test.go#L518

File: cmd/replay/espresso_validation.go (hunk 1 above)
Comment (zacshowa) [reply cmd/replay/espresso_validation.go:27]: I'm a little confused what you mean by "Not all messages in the espresso mode can be called `EspressoMessage`"
I'm aware that we want to allow other messages through when the hotshot height is zero, but if `IsL2NonEspressoMsg` isn't enough to determine the set of messages we want to allow through, what else do I need to look at?

Do we not need to be regularly querying the VM to get the height? It does seem like querying the VM every time could add overhead to the process. Should I just be looking for the height in whatever messages we have access to?

Comment (ImJeremyHe) [reply cmd/replay/espresso_validation.go:27]: https://github.com/EspressoSystems/nitro-espresso-integration/blob/integration/arbos/parse_l2.go#L506
```rs
func IsEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
	return msg.Header.Kind == arbostypes.L1MessageType_L2Message &&
		msg.L2msg[0] == L2MessageKind_EspressoTx
}

func IsL2NonEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
	return msg.Header.Kind == arbostypes.L1MessageType_L2Message &&
		msg.L2msg[0] != L2MessageKind_EspressoTx
}
```
The code below is not the same as the condition it was.
```
isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
validatingEspressoLivenessFailure := isNonEspressoMessage && (isEnabled || hotShotHeight != 0)
```

Comment (ImJeremyHe) [reply cmd/replay/espresso_validation.go:27]:
```
validatingEspressoLivenessFailure := arbos.IsL2NonEspressoMsg(message.Message) && chainConfig.ArbitrumChainParams.EnableEspresso
```

Comment (zacshowa) [reply cmd/replay/espresso_validation.go:27]: The reason that the condition isn't the same is to add the check referenced in the ticket such that the validator panics when the incoming message is not an espresso message, and the hotshot height is non zero.

Comment (ImJeremyHe) [reply cmd/replay/espresso_validation.go:27]:
```rs
	isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
	isEspressoMessage := !isNonEspressoMessage
```
is not correct. Sorry for the misleading code or interpretion.

Actually this is how we tell if the message is `EspressoMessage` or not:
```rs
func IsEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
    return msg.Header.Kind == arbostypes.L1MessageType_L2Message && msg.L2msg[0] == L2MessageKind_EspressoTx }
```
and it is not just `!isL2NonEspressoMessage()`.

There are many kinds of messages in the nitro system. The most important one that we need to pay attention to is the `L2Message` with the specific l2message type `EspressoTx` (that is what we called `EspressoMessage`). If the hotshot is good, the l2 messages will have the one type. `EspressoTx`.

When checking liveness failure, we check:
- if nitro are receiving the l2 message that is not `EspressoTx`
- if `espresso` is enabled.

**Comment by zacshowa** [comment]
@ImJeremyHe I feel it would be pertinent to double check my logic here. This change intends to keep the previous panic cases covered, as well as handle the new case where the message is not an espresso message, and the HotShot height is non-zero.
From my understanding, the previous behavior was to only have the VM panic when we fail the `validatingAgainstEspresso` check, and the error case occurs. If this is the expected and correct behavior, this PR should maintain that behavior.

File: cmd/replay/espresso_validation.go [hunk 2]
```diff
@@ -2,39 +2,48 @@ package main
...
+		// something incorrect has occurred and we must panic
+		return validatingAgainstEspresso, func() {
+			panic("The messaged received by the STF is not an Espresso message, but the validator is running in Espresso mode")
+		}
+	}
+	panicCase := func() {
+		// This no-op allows us to cause no panic in the code that receives the result of this function while keeping the function signature consistent.
 	}
-	return validatingAgainstEspresso
+	return validatingAgainstEspresso, panicCase
```
Comment (ImJeremyHe) [inline cmd/replay/espresso_validation.go:48]: We can directly return `nil` here

**Review by ImJeremyHe — APPROVED** [review]
For now, the e2e test is so complicated and as long as the tests pass, it should be ok. Many servers/components are running and they can not be stopped at the same time, so I think it is normal that the some error logs occur

**Pull request merged**